### PR TITLE
Generalize `open_long` & `open_short` test cases into `InstanceTest` suite

### LINKED ISSUE #885: Generalize `open_long` instance test cases
Build on #883 and add testing coverage for `open_long` test cases.

- add `test_open_long_with_base` to `InstanceTestingSuite` contract.
- add `test_open_long_with_shares` to `InstanceTestingSuite` contract.

### PR DESCRIPTION
**Pull request by cashd**
# Resolved Issues
Resolves #885 #890 

# Description
Build on #883 and adds generalized testing coverage for open_long test cases.

- adds `test_open_long_with_base` to `InstanceTest` contract.
- adds `test_open_long_with_shares` to `InstanceTest` contract.
- adds `test_open_short_with_base` to `InstanceTest` contract.
- adds `test_open_short_with_shares` to `InstanceTest` contract.

# Review Checklists

Please check each item **before approving** the pull request. While going
through the checklist, it is recommended to leave comments on items that are
referenced in the checklist to make sure that they are reviewed. If there are
multiple reviewers, copy the checklists into sections titled `## [Reviewer Name]`.
If the PR doesn't touch Solidity and/or Rust, the corresponding checklist can
be removed.

### Solidity

- [ ] **Tokens**
    - [ ] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [ ] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [ ] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [ ] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [ ] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [ ] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [ ] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### Rust

- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [ ] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?

### CODE REVIEW

File: test/utils/InstanceTest.sol [hunk 1]
```diff
@@ -458,4 +496,128 @@ abstract contract InstanceTest is HyperdriveTest {
...
+
+    /// @dev Fuzz Test to ensure deposit accounting is correct for opening longs
+    ///      with the share token.
+    /// @param basePaid Amount in terms of base to open a long.
+    function test_open_long_with_shares(uint256 basePaid) external {
+        // Get balance information before opening a long.
+        (
+            uint256 totalBaseSupplyBefore,
+            uint256 totalSharesSupplyBefore
+        ) = getSupply();
```
Comment (jalextowle) [inline test/utils/InstanceTest.sol:508]: This looks like the wrong order (above `shares` come before `base`). FWIW, this order is more intuitive to me (`base` and then `shares`), but regardless it looks like this needs to be updated.

Comment (jalextowle) [reply test/utils/InstanceTest.sol:508]: Why did this and the below test pass with this order?

File: test/utils/InstanceTest.sol [hunk 2]
```diff
@@ -244,8 +244,7 @@ abstract contract InstanceTest is HyperdriveTest {
         }
     }
 
-    /// @dev Deploys the Hyperdrive Factory contract and
-    ///      sets the default pool configuration.
```
Comment (jalextowle) [inline test/utils/InstanceTest.sol:248]: This comment was shorter to stay within 80 chars. We'll want to do this everywhere to stay consistent.

File: test/utils/InstanceTest.sol [hunk 3]
```diff
@@ -458,4 +496,128 @@ abstract contract InstanceTest is HyperdriveTest {
...
+            vm.expectRevert(
+                isBaseETH
+                    ? IHyperdrive.NotPayable.selector
+                    : IHyperdrive.UnsupportedToken.selector
+            );
+        }
+
+        // Bob opens a long by depositing the base token.
+        vm.startPrank(bob);
+        hyperdrive.openLong{ value: isBaseETH ? basePaid : 0 }(
```
Comment (jalextowle) [inline test/utils/InstanceTest.sol:578]: It would be good to check the return value.

File: test/utils/InstanceTest.sol [hunk 4]
```diff
@@ -458,4 +496,128 @@ abstract contract InstanceTest is HyperdriveTest {
...
+        basePaid = basePaid.normalizeToRange(
+            2 * hyperdrive.getPoolConfig().minimumTransactionAmount,
+            maxLongAmount > maxShareAmount ? maxShareAmount : maxLongAmount
+        );
+
+        // Convert the amount to deposit in terms of the share token.
+        uint256 sharesPaid = convertToShares(basePaid);
+
+        // Bob opens a long by depositing the share token.
+        openLong(bob, sharesPaid, false);
```
Comment (jalextowle) [inline test/utils/InstanceTest.sol:530]: Same here, it would be good to check the return.

File: test/instances/steth/StETHHyperdrive.t.sol [hunk 5]
```diff
@@ -92,6 +92,100 @@ contract StETHHyperdriveTest is InstanceTest {
...
+
+    /// @dev Fetches the token balance information of an account.
+    function getTokenBalances(
+        address account
+    ) internal view override returns (uint256, uint256) {
+        return (LIDO.sharesOf(account), LIDO.balanceOf(account));
+    }
+
+    /// @dev Verifies that deposit accounting is correct when opening positions.
+    function verifyDeposit(
```
Comment (jalextowle) [inline test/instances/steth/StETHHyperdrive.t.sol:108]: Any reason not to abstract this further? It seems like we could define some functions like `getSupply` and `getTokenBalances` that we could use to define this. Maybe I'm missing something, but it seems like the kind of thing that we shouldn't need to implement from scratch for each yield source.

File: test/utils/InstanceTest.sol (hunk 1 above)
Comment (cashd) [reply test/utils/InstanceTest.sol:508]: Agreed on base then shares, I confused my own self haha. This is actually how I implemented the override functions so the test cases are passing.

File: test/utils/InstanceTest.sol (hunk 2 above)
Comment (cashd) [reply test/utils/InstanceTest.sol:248]: Is the style requirement only comment part or entire line (including natspec)? Do you use a plugin to automatically do this for you?

File: test/instances/steth/StETHHyperdrive.t.sol (hunk 5 above)
Comment (cashd) [reply test/instances/steth/StETHHyperdrive.t.sol:108]: I've considered this also, but there is some divergence between what values we assert depending on the instance. The biggest difference is with stETH and the other instances. For example stETH is checking the underlying balance and shares when depositing as shares. With yield bearing tokens it's just a share token balance check. Given the divergence I plan to tackle the lowest hanging fruit first, open/close positions test cases.

**Comment by cashd** [comment]
I added `open_short` test cases to this PR

File: test/utils/InstanceTest.sol [hunk 6]
```diff
@@ -315,12 +315,59 @@ abstract contract InstanceTest is HyperdriveTest {
...
+    /// @param account The account to fetch token balances of.
+    /// @return sharesBalance The shares token balance of the account.
+    /// @return baseBalance The base token balance of the account.
+    function getTokenBalances(
+        address account
+    )
+        internal
+        view
+        virtual
+        returns (uint256 sharesBalance, uint256 baseBalance);
```
Comment (jalextowle) [inline test/utils/InstanceTest.sol:361]: Should this be base and then shares? That would be more consistent with the below function.

File: test/utils/InstanceTest.sol [hunk 7]
```diff
@@ -458,4 +505,340 @@ abstract contract InstanceTest is HyperdriveTest {
...
+        // Ensure that Bob received the correct amount of bonds.
+        assertEq(
+            hyperdrive.balanceOf(
+                AssetId.encodeAssetId(AssetId.AssetIdPrefix.Long, maturityTime),
+                bob
+            ),
+            bondAmount
+        );
+
+        // If base deposits are supported we ensure the deposit accounting is correct.
```
Comment (jalextowle) [inline test/utils/InstanceTest.sol:642]: Nit:

```suggestion
        // If base deposits are supported, we ensure the deposit accounting is correct.
```

File: test/utils/InstanceTest.sol (hunk 2 above)
Comment (jalextowle) [reply test/utils/InstanceTest.sol:248]: It's all of it. I don't have a plugin do it, but I do have a marker in my editor that tells me where 80 characters is. A plugin would be great.

File: test/instances/steth/StETHHyperdrive.t.sol (hunk 5 above)
Comment (jalextowle) [reply test/instances/steth/StETHHyperdrive.t.sol:116]: Ok, that makes sense. As long as we get to this eventually, I'm happy.

**Review by jalextowle — APPROVED** [review]
LGTM once the last few comments are addressed

**Pull request merged**